controllers/osbuildconfig_controller.go

```diff
@@ -70,15 +81,139 @@ func (r *OSBuildConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, nil
 	}
 
-	err = r.OSBuildCRCreator.Create(ctx, osBuildConfig)
+	newOSBuildInstanceIsNeeded, err := r.checkIfNewOSBuildInstanceIsNeeded(ctx, osBuildConfig, logger)
+	if err != nil {
+		return ctrl.Result{Requeue: true, RequeueAfter: RequeueForShortDuration}, nil
+	}
+
+	if newOSBuildInstanceIsNeeded {
+		// create new OSBuild instance
+		targetBuild := osBuildConfig.Spec.Details.TargetImage.TargetImageType
+		if targetBuild == osbuilderv1alpha1.EdgeInstallerImageType {
+			// start with edge-container OSBuild instance
+			targetBuild = osbuilderv1alpha1.EdgeContainerImageType
+		}
+		return r.createOSBuildInstance(ctx, logger, osBuildConfig, targetBuild)
+	} else {
+		return r.updateOSBuildConfigCurrentStatus(ctx, logger, osBuildConfig)
+	}
+}
+
+func (r *OSBuildConfigReconciler) checkIfNewOSBuildInstanceIsNeeded(ctx context.Context, osBuildConfig *osbuilderv1alpha1.OSBuildConfig, logger logr.Logger) (bool, error) {
+	userConfiguration := osbuilderv1alpha1.UserConfiguration{}
+	// TODO: sort before
+	if osBuildConfig.Spec.Details.Customizations != nil {
+		userConfiguration.Customizations = osBuildConfig.Spec.Details.Customizations
+	}
+	if osBuildConfig.Spec.Template != nil {
+		userConfiguration.Template = osBuildConfig.Spec.Template
+	}
+
+	if osBuildConfig.Status.LastKnownUserConfiguration == nil || !cmp.Equal(osBuildConfig.Status.LastKnownUserConfiguration, userConfiguration) {
+		patch := client.MergeFrom(osBuildConfig.DeepCopy())
+		osBuildConfig.Status.LastKnownUserConfiguration = &userConfiguration
+		errPatch := r.OSBuildConfigRepository.PatchStatus(ctx, osBuildConfig, &patch)
+		if errPatch != nil {
+			logger.Error(errPatch, "Failed to patch OSBuildConfig status")
+			return false, errPatch
+		}
+
+		return true, nil
+	}
+
+	if osBuildConfig.Annotations == nil {
+		return false, nil
+	}
+
+	webhookTriggerTS, ok := osBuildConfig.Annotations[webHookAnnotationKey]
+	if !ok {
+		return false, nil
+	}
+
+	if osBuildConfig.Status.LastWebhookTriggerTS == "" || webhookTriggerTS != osBuildConfig.Status.LastWebhookTriggerTS {
+		patch := client.MergeFrom(osBuildConfig.DeepCopy())
+		osBuildConfig.Status.LastWebhookTriggerTS = webhookTriggerTS
+		errPatch := r.OSBuildConfigRepository.PatchStatus(ctx, osBuildConfig, &patch)
+		if errPatch != nil {
+			logger.Error(errPatch, "Failed to patch OSBuildConfig status")
+			return false, errPatch
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (r *OSBuildConfigReconciler) updateOSBuildConfigCurrentStatus(ctx context.Context, logger logr.Logger, osBuildConfig *osbuilderv1alpha1.OSBuildConfig) (ctrl.Result, error) {
+	osBuildName := fmt.Sprintf("%s-%d", osBuildConfig.Name, *osBuildConfig.Status.LastVersion)
+	osBuild, err := r.OSBuildRepository.Read(ctx, osBuildName, osBuildConfig.Namespace)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			return ctrl.Result{RequeueAfter: 30 * time.Second}, err
+			// should create a new one from the same target image type
```

Comment:
I don't understand this `if`. Why should this happen and why is OK that it happened?

Reply:
For example - 
1- Create osbuildConfig
2- update it immediately so the OSBuild instance wasn't created yet 
in that case, I think we would like to create a new instance

Reply:
If the controller ran between stages 1 and 2, then it will create an OSBuild instance and then run again after stage 2 and create another OSBuild instance.
If it ran after stage 2, it will create an OSBuild instance from the top of the configuration.

I don't see a scenario in which the instance does not exist. If we get that, I think we're in an error state. Please keep in mind that OSBuild instances are not deleted.

Reply:
So what we should do in that scenario? requeue for short duration? or Done?

Reply:
I think we should treat it like any other (i.e. not check the error type)